Lib/unittest/test/testmock/testmock.py

```diff
@@ -1336,6 +1336,54 @@ def test_assert_has_calls(self):
                         )
 
 
+    def test_assert_has_calls_nested_spec(self):
+        class Something:
+
+            def __init__(self): pass
+            def meth(self, a, b, c, d=None): pass
+
+            class Foo:
+
+                def __init__(self, a): pass
+                def meth1(self, a, b): pass
+
+        mock_class = create_autospec(Something)
+
+        for m in [mock_class, mock_class()]:
+            m.meth(1, 2, 3, d=1)
```

Comment:
I'm confused: for the `mock_class` case, why is this expected to work? `meth` isn't a classmethod...

Reply:
I too had the same confusion :) It seems for both class and instance the signature skips self as in the below example. There is some note in docs at https://docs.python.org/3/library/unittest.mock.html#create-autospec but I am also little surprised at this.

> If a class is used as a spec then the return value of the mock (the instance of the class) will have the same spec. You can use a class as the spec for an instance object by passing instance=True. The returned mock will only be callable if instances of the mock are callable.

```python
from unittest.mock import *
import inspect

class A:

    def meth1(self, a): pass

a = create_autospec(A)
a.meth1(1)
print(inspect.signature(a.meth1))
print(a.mock_calls)

# Use as an object
obj = a()
obj.meth1(1)
print(inspect.signature(obj.meth1))
print(obj.mock_calls)
```

```
./python.exe /tmp/bar.py
(a)
[call.meth1(1)]
(a)
[call.meth1(1)]
```

Similar test : 

https://github.com/python/cpython/blob/0f6f73ff3c2e46144884c4d418e67939a042aa23/Lib/unittest/test/testmock/testhelpers.py#L533-L543

Reply:
It's unrelated to this PR, but this just feels like a bug to me. If you've autospecced from a class, the mock shouldn't have things that behave like classmethods if those things aren't classmethods on the object the spec is from.

The "similar test" you've included is just as confusing: why should the mock have `one`, `two` or `three` attributes when the original does not? @voidspace?

Reply:
one, two and three are methods of `SomeClass` that is inherited by `Sub`

Reply:
Ah! That makes more sense :-)